Mathlib/Combinatorics/Additive/ErdosGinzburgZiv.lean

```diff
@@ -0,0 +1,185 @@
+      (𝒜 : Set (Finset ι)).Pairwise _root_.Disjoint ∧
+        ∀ ⦃t⦄, t ∈ 𝒜 → t ⊆ s ∧ t.card = n ∧ ↑n ∣ ∑ i ∈ t, a i by
+     -- Assume `𝒜` is a family of `2 * m - 1` sets, each of size `n` and sum divisible by `n`.
+      obtain ⟨𝒜, h𝒜card, h𝒜disj, h𝒜⟩ := this _ le_rfl
+      -- By induction hypothesis on `m`, find a subfamily `ℬ` of size `m` such that the sum over
+      -- `t ∈ ℬ` of `(∑ i ∈ t, a i) / n` is divisible by `m`.
+      obtain ⟨ℬ, hℬ𝒜, hℬcard, hℬ⟩ := ihm (fun t ↦ (∑ i ∈ t, a i) / n) h𝒜card.ge
+      -- We are done.
+      refine ⟨ℬ.biUnion fun x ↦ x, biUnion_subset.2 fun t ht ↦ (h𝒜 $ hℬ𝒜 ht).1, ?_, ?_⟩
+      · rw [card_biUnion (h𝒜disj.mono hℬ𝒜), sum_const_nat fun t ht ↦ (h𝒜 $ hℬ𝒜 ht).2.1, hℬcard]
+      rwa [sum_biUnion, natCast_mul, mul_comm, ← Int.dvd_div_iff_mul_dvd, Int.sum_div]
+      · exact fun t ht ↦ (h𝒜 $ hℬ𝒜 ht).2.2
+      · exact dvd_sum fun t ht ↦ (h𝒜 $ hℬ𝒜 ht).2.2
+      · exact h𝒜disj.mono hℬ𝒜
+    -- Now, let's find those `2 * m - 1` sets.
+    rintro k hk
+    -- We induct on the size `k ≤ 2 * m - 1` of the family we are constructing.
+    induction' k with k ih
+    -- For `k = 0`, the empty family trivially works.
+    · exact ⟨∅, by simp⟩
+    -- At `k + 1`, call `𝒜` the existing family of size `k ≤ 2 * m - 2`.
+    obtain ⟨𝒜, h𝒜card, h𝒜disj, h𝒜⟩ := ih (Nat.le_of_succ_le hk)
+    -- There are at least `2 * (m * n) - 1 - k * n ≥ 2 * m - 1` elements in `s` that have not been
+    -- taken in any element of `𝒜`.
+    have : 2 * n - 1 ≤ (s \ 𝒜.biUnion id).card := by
+      calc
+        _ ≤ (2 * m - k) * n - 1 := by gcongr; omega
+        _ = (2 * (m * n) - 1) - ∑ t ∈ 𝒜, t.card := by
+          rw [tsub_mul, mul_assoc, tsub_right_comm, sum_const_nat fun t ht ↦ (h𝒜 ht).2.1, h𝒜card]
+        _ ≤ s.card - (𝒜.biUnion id).card := by gcongr; exact card_biUnion_le
+        _ ≤ (s \ 𝒜.biUnion id).card := le_card_sdiff ..
+    -- So by the induction hypothesis on `n` we can find a new set `t` of size `n` and sum divisible
+    -- by `n`.
+    obtain ⟨t₀, ht₀, ht₀card, ht₀sum⟩ := ihn a this
+    -- This set is distinct and disjoint from the previous ones, so we are done.
+    have : t₀ ∉ 𝒜 := by
+      rintro h
+      obtain rfl : n = 0 := by
+        simpa [← card_eq_zero, ht₀card] using sdiff_disjoint.mono ht₀ $ subset_biUnion_of_mem id h
+      omega
+    refine ⟨𝒜.cons t₀ this, by rw [card_cons, h𝒜card], ?_, ?_⟩
+    · simp only [cons_eq_insert, coe_insert, Set.pairwise_insert_of_symmetric symmetric_disjoint,
+        mem_coe, ne_eq]
+      exact ⟨h𝒜disj, fun t ht _ ↦ sdiff_disjoint.mono ht₀ $ subset_biUnion_of_mem id ht⟩
+    · simp only [cons_eq_insert, mem_insert, forall_eq_or_imp, and_assoc]
+      exact ⟨ht₀.trans sdiff_subset, ht₀card, ht₀sum, h𝒜⟩
+
+/-- The **Erdős–Ginzburg–Ziv theorem** for `ℤ/nℤ`.
+
+Any sequence of at least `2 * n - 1` elements of `ZMod n` contains a subsequence of `n` elements
+whose sum is zero. -/
+theorem ZMod.erdos_ginzburg_ziv (a : ι → ZMod n) (hs : 2 * n - 1 ≤ s.card) :
```

Comment:
Since the informal versions use the term 'sequence', let's also have a list/multiset version (of course derived from these).

Reply:
Does the informal "sequence" really mean the formal `Multiset`?

Reply:
That's why I offered List instead :) In any case, my point is that there are other formulations of this statement, and we might as well provide them too.

Reply:
Yeah but sometimes it's better to force people to use the superior statement. I've added the multiset versions but I'm unsure they are useful